test/micro/org/openjdk/bench/java/lang/foreign/InternalStrLen.java

```diff
@@ -0,0 +1,161 @@
+ */
+package org.openjdk.bench.java.lang.foreign;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.ValueLayout;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.lang.foreign.ValueLayout.*;
+import static jdk.internal.foreign.StringSupport.*;
+
+@BenchmarkMode(Mode.AverageTime)
+@Warmup(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@State(Scope.Benchmark)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(value = 3, jvmArgsAppend = {"--add-exports=java.base/jdk.internal.foreign=ALL-UNNAMED", "--enable-native-access=ALL-UNNAMED", "--enable-preview"})
+public class InternalStrLen {
+
+    private MemorySegment singleByteSegment;
+    private MemorySegment singleByteSegmentMisaligned;
+    private MemorySegment doubleByteSegment;
+    private MemorySegment quadByteSegment;
+
+    @Param({"1", "4", "16", "251", "1024"})
+    int size;
+
+    @Setup
+    public void setup() {
+        var arena = Arena.ofAuto();
+        singleByteSegment = arena.allocate((size + 1L) * Byte.BYTES);
+        singleByteSegmentMisaligned = arena.allocate((size + 1L) * Byte.BYTES);
+        doubleByteSegment = arena.allocate((size + 1L) * Short.BYTES);
+        quadByteSegment = arena.allocate((size + 1L) * Integer.BYTES);
+        Stream.of(singleByteSegment, doubleByteSegment, quadByteSegment)
+                .forEach(s -> IntStream.range(0, (int) s.byteSize() - 1)
+                        .forEach(i -> s.set(
+                                ValueLayout.JAVA_BYTE,
+                                i,
+                                (byte) ThreadLocalRandom.current().nextInt(1, 254)
+                        )));
+        singleByteSegment.set(ValueLayout.JAVA_BYTE, singleByteSegment.byteSize() - Byte.BYTES, (byte) 0);
+        doubleByteSegment.set(ValueLayout.JAVA_SHORT, doubleByteSegment.byteSize() - Short.BYTES, (short) 0);
+        quadByteSegment.set(ValueLayout.JAVA_INT, quadByteSegment.byteSize() - Integer.BYTES, 0);
+        singleByteSegmentMisaligned = arena.allocate(singleByteSegment.byteSize() + 1).
+                asSlice(1);
+        MemorySegment.copy(singleByteSegment, 0, singleByteSegmentMisaligned, 0, singleByteSegment.byteSize());
+    }
+
+    @Benchmark
+    public int legacySingleByte() {
```

Comment:
I think we should rename `legacy` and `new` to something else. E.g. `byte` vs. `bulk`. You might rename the various methods as `byte_2` (meaning byte-by-byte, with two bytes chars).